javaharness/java/arcs/android/AndroidArcsEnvironment.java

```diff
@@ -0,0 +1,172 @@
+@Singleton
+final class AndroidArcsEnvironment {
+
+  interface ReadyListener {
+    void onReady(List<String> recipes);
+  }
+
+  private static final String FIELD_MESSAGE = "message";
+  private static final String MESSAGE_READY = "ready";
+  private static final String FIELD_READY_RECIPES = "recipes";
+  private static final String MESSAGE_DATA = "data";
+  private static final String MESSAGE_OUTPUT = "output";
+  private static final String MESSAGE_PEC = "pec";
+  private static final String FIELD_TRANSACTION_ID = "tid";
+  private static final String FIELD_DATA = "data";
+  private static final String FIELD_PEC_ID = "id";
+  private static final String FIELD_SESSION_ID = "sessionId";
+
+  private final List<ReadyListener> readyListeners = new ArrayList<>();
+  private final PortableJsonParser jsonParser;
+  private final PECPortManager pecPortManager;
+  private final UiBroker uiBroker;
+  private final ShellApi shellApi;
+  private final Handler uiThreadHandler;
+  private WebView webView;
+  // Fetches the up-to-date properties on every get().
+  private Provider<RuntimeSettings> runtimeSettings;
+
+  @Inject
+  public AndroidArcsEnvironment(
+    PortableJsonParser jsonParser,
+    PECPortManager pecPortManager,
+    UiBroker uiBroker,
+    ShellApi shellApi,
+    Provider<RuntimeSettings> runtimeSettings) {
+    this.jsonParser = jsonParser;
+    this.pecPortManager = pecPortManager;
+    this.uiBroker = uiBroker;
+    this.shellApi = shellApi;
+    this.runtimeSettings = runtimeSettings;
+
+    this.uiThreadHandler = new Handler(Looper.getMainLooper());
+  }
+
+  public void addReadyListener(ReadyListener listener) {
+    readyListeners.add(listener);
+  }
+
+  public void fireReadyEvent(List<String> recipes) {
+    readyListeners.forEach(listener -> listener.onReady(recipes));
+  }
+
+  public void init(Context context) {
+    webView = new WebView(context);
+    webView.setVisibility(View.GONE);
+    webView.getSettings().setAppCacheEnabled(false);
+    webView.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
+    webView.clearCache(true);
+    webView.setWebContentsDebuggingEnabled(true);
+
+    WebSettings arcsSettings = webView.getSettings();
+    arcsSettings.setDatabaseEnabled(true);
+    arcsSettings.setGeolocationEnabled(true);
+    arcsSettings.setJavaScriptEnabled(true);
+    arcsSettings.setDomStorageEnabled(true);
+    arcsSettings.setSafeBrowsingEnabled(false);
+    // These two are needed for file:// URLs to work and load subresources
+    arcsSettings.setAllowFileAccessFromFileURLs(true);
+    // needed to allow WebWorkers to work in FileURLs.
+    arcsSettings.setAllowUniversalAccessFromFileURLs(true);
+
+    webView.setWebViewClient(new WebViewClient() {
+      @Override
+      public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
+        return super.shouldInterceptRequest(view, request);
+      }
+    });
+
+    webView.addJavascriptInterface(this, "DeviceClient");
+
```

Comment:
The only downside of this is that it makes it harder to rip-out the WebView. In the old syste, DeviceClient, like ShellApi, was injected, and the Web implementation was hidden behind it. This means that mock versions, or particle accelerator versions that handled the JSON messaging, but did not have a WebView would be possible. 

The way this is now, you'd have to fork the class. I'd at least move all of the non-WebView logic to a base class, and then make this class extend ti with the WebView specific setup.

Reply:
I actually felt the device client logic is tied up to the WebView+PipesShell implementation as its interface relies on receiving messages with certain formats from PipesShell to work. I'm not sure if this is still the case for accelerator or other implementations? Seems to me these alter environments are still in design states so I tend to not over-design for it. We can always refactor this part again once we have a more consolidated design/impl for alter environments. 

Right now I'd like to keep it simple as possible.